File: admin/src/test/java/com/beat/admin/promotion/facade/AdminPromotionFacadeTest.java
package com.beat.admin.promotion.facade;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;

import com.beat.admin.promotion.api.request.CarouselHandleRequest;
import com.beat.admin.promotion.application.command.CarouselHandleCommand;
import com.beat.admin.promotion.application.command.AdminPromotionCommandService;
import com.beat.admin.promotion.application.result.AdminPromotionPresignedUrlResults.BannerPresignedUrlResult;
import com.beat.admin.promotion.application.result.AdminPromotionPresignedUrlResults.CarouselPresignedUrlsResult;
import com.beat.admin.promotion.application.result.AdminPromotionResults;
import com.beat.admin.promotion.application.query.AdminPromotionQueryService;

class AdminPromotionFacadeTest {

	@Test
	void facadeDelegatesPromotionScenariosToApplicationServices() {
		AdminPromotionQueryService queryService = mock(AdminPromotionQueryService.class);
		AdminPromotionCommandService commandService = mock(AdminPromotionCommandService.class);
		AdminPromotionFacade adminPromotionFacade = new AdminPromotionFacade(queryService, commandService);
		CarouselHandleRequest request = new CarouselHandleRequest(List.of());
		when(queryService.issueAllPresignedUrlsForCarousel(1L, List.of("carousel.png")))
			.thenReturn(new CarouselPresignedUrlsResult(Map.of()));
		when(queryService.issuePresignedUrlForBanner(1L, "banner.png"))
			.thenReturn(new BannerPresignedUrlResult("banner-url"));
		when(queryService.findAllPromotionsSortedByCarouselNumber(1L))
			.thenReturn(new AdminPromotionResults(List.of()));
		when(commandService.processAllPromotionsSortedByCarouselNumber(1L, new CarouselHandleCommand(List.of())))
			.thenReturn(new AdminPromotionResults(List.of()));

		adminPromotionFacade.checkMemberAndIssueAllPresignedUrlsForCarousel(1L, List.of("carousel.png"));
		adminPromotionFacade.checkMemberAndIssuePresignedUrlForBanner(1L, "banner.png");
		adminPromotionFacade.checkMemberAndFindAllPromotionsSortedByCarouselNumber(1L);
		adminPromotionFacade.checkMemberAndProcessAllPromotionsSortedByCarouselNumber(1L, request);

		verify(queryService).issueAllPresignedUrlsForCarousel(1L, List.of("carousel.png"));
		verify(queryService).issuePresignedUrlForBanner(1L, "banner.png");
		verify(queryService).findAllPromotionsSortedByCarouselNumber(1L);
		verify(commandService).processAllPromotionsSortedByCarouselNumber(1L, new CarouselHandleCommand(List.of()));
	}
}

File: admin/src/test/kotlin/com/beat/admin/promotion/facade/AdminPromotionFacadeTest.kt
package com.beat.admin.promotion.facade

import com.beat.admin.promotion.api.request.CarouselHandleRequest
import com.beat.admin.promotion.application.command.AdminPromotionCommandService
import com.beat.admin.promotion.application.command.CarouselHandleCommand
import com.beat.admin.promotion.application.query.AdminPromotionQueryService
import com.beat.admin.promotion.application.result.AdminPromotionPresignedUrlResults.BannerPresignedUrlResult
import com.beat.admin.promotion.application.result.AdminPromotionPresignedUrlResults.CarouselPresignedUrlsResult
import com.beat.admin.promotion.application.result.AdminPromotionResults
import org.junit.jupiter.api.Test
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify

class AdminPromotionFacadeTest {

    @Test
    fun facadeDelegatesPromotionScenariosToApplicationServices() {
        val queryService = mock(AdminPromotionQueryService::class.java)
        val commandService = mock(AdminPromotionCommandService::class.java)
        val adminPromotionFacade = AdminPromotionFacade(queryService, commandService)
        val request = CarouselHandleRequest(emptyList())
        `when`(queryService.issueAllPresignedUrlsForCarousel(1L, listOf("carousel.png")))
            .thenReturn(CarouselPresignedUrlsResult(emptyMap()))
        `when`(queryService.issuePresignedUrlForBanner(1L, "banner.png"))
            .thenReturn(BannerPresignedUrlResult("banner-url", "prod/banner/banner.png"))
        `when`(queryService.findAllPromotionsSortedByCarouselNumber(1L))
            .thenReturn(AdminPromotionResults(emptyList()))
        `when`(commandService.processAllPromotionsSortedByCarouselNumber(1L, CarouselHandleCommand.from(emptyList())))
            .thenReturn(AdminPromotionResults(emptyList()))

        adminPromotionFacade.checkMemberAndIssueAllPresignedUrlsForCarousel(1L, listOf("carousel.png"))
        adminPromotionFacade.checkMemberAndIssuePresignedUrlForBanner(1L, "banner.png")
        adminPromotionFacade.checkMemberAndFindAllPromotionsSortedByCarouselNumber(1L)
        adminPromotionFacade.checkMemberAndProcessAllPromotionsSortedByCarouselNumber(1L, request)

        verify(queryService).issueAllPresignedUrlsForCarousel(1L, listOf("carousel.png"))
        verify(queryService).issuePresignedUrlForBanner(1L, "banner.png")
        verify(queryService).findAllPromotionsSortedByCarouselNumber(1L)
        verify(commandService).processAllPromotionsSortedByCarouselNumber(1L, CarouselHandleCommand.from(emptyList()))
    }
}